bin/find-fixable-orphans.py

```diff
@@ -0,0 +1,293 @@
+
+
+# Get all keys from AWS S3
+def get_s3_bucket_raw_keys_list(s3_bucket=AWS_S3_BUCKET):
+    print(f"Getting raw AWS S3 keys recursively ({s3_bucket}) ...", file=sys.stderr)
+
+    s3_resource = boto3.resource("s3")
+    s3_resource_bucket = s3_resource.Bucket(s3_bucket)
+    this_s3_bucket_raw_keys_list = [
+        s3_object.key for s3_object in s3_resource_bucket.objects.all()
+    ]
+
+    print(
+        f"{len(this_s3_bucket_raw_keys_list)} rows retrieved",
+        file=sys.stderr,
+    )
+
+    return this_s3_bucket_raw_keys_list
+
+
+# Get the keys present and absent across NZSL Signbank and S3, to dictionary
+def create_all_keys_dict(this_nzsl_raw_keys_dict, this_s3_bucket_raw_keys_list):
+    print(
+        "Getting keys present and absent across NZSL Signbank and S3 ...",
+        file=sys.stderr,
+    )
+    this_all_keys_dict = {}
+
+    # Find S3 keys that are present in NZSL, or absent
+    for video_key in this_s3_bucket_raw_keys_list:
+        dict_row = this_nzsl_raw_keys_dict.get(video_key, None)
+        if dict_row:
+            # NZSL glossvideo record for this S3 key
+            this_all_keys_dict[video_key] = [
+                True,  # NZSL PRESENT
+                True,  # S3 PRESENT
+            ] + dict_row
+        else:
+            # S3 key with no corresponding NZSL glossvideo record
+            this_all_keys_dict[video_key] = [
+                False,  # NZSL Absent
+                True,  # S3 PRESENT
+            ] + [""] * 6
+
+    # Find NZSL keys that are absent from S3 (present in both handled above)
+    for video_key, dict_row in this_nzsl_raw_keys_dict.items():
+        if video_key not in this_s3_bucket_raw_keys_list:
+            # gloss/glossvideo record with no corresponding S3 key
+            # Either:
+            # video_key is real, but the S3 object is missing
+            # video_key is fake (to handle the FULL JOIN) and this gloss/glossvideo never had an S3 object
+            this_all_keys_dict[video_key] = [
+                True,  # NZSL PRESENT
+                False,  # S3 Absent
+            ] + dict_row
+
+    return this_all_keys_dict
+
+
+def find_orphans():
+    all_keys_dict = create_all_keys_dict(
+        get_nzsl_raw_keys_dict(), get_s3_bucket_raw_keys_list()
+    )
+    print("Finding fixable orphans", file=sys.stderr)
+
+    print(CSV_DELIMITER.join(GLOBAL_COLUMN_HEADINGS))
+
+    # Traverse all the NZSL Signbank glosses that are missing S3 objects
+    for video_key, [
+        key_in_nzsl,
+        key_in_s3,
+        gloss_idgloss,
+        gloss_created_at,
+        gloss_id,
+        video_id,
+        gloss_public,
+        video_public,
+    ] in all_keys_dict.items():
+
+        if not key_in_nzsl:
+            # This is an S3 object, not a Signbank record
+            continue
+
+        if key_in_s3:
+            # This Signbank record already has an S3 object, all is well
+            continue
+
+        # Business rule
+        if int(gloss_id) < 8000:
```

Comment:
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent.  -->
**suggestion:** Exclude these in the SQL

If we don't care about gloss ids that are greater than 8000 - then excluding them at source (the SQL query) then you have less data to deal with. ‏

Reply:
Do we actually still need this in the first place? my expectation is in the long-run these should all go away, so now that we've done the running of these scripts, we can remove this as they should be sorted by the time we next run these.

Otherwise, can you write up a proper comment explaining what the business rule actually is? a good test you can use to confirm if you're captured enough detail is to ask "could someone determine if the business rule was still relevant based on just the contents of this script?"

Reply:
I agree with the SQL comment, but I agree that this rule can go away now. It was a patch-over for a legacy problem, one that no longer exists. The only way it could come back is if somebody did another import from the old Freelex database (long story, Jon) and we've pretty much knocked that on the head as a possibility.
I'll just remove this line.  When I do the documentation properly I'll copy some of the correspondence with Micky about <8000 into it.